components/playlist/browser/playlist_service.cc

```diff
@@ -0,0 +1,575 @@
+  audio_media_file_downloader_->RequestCancelCurrentPlaylistGeneration();
+  url_loaders_.clear();
+  base::queue<base::Value> empty_queue;
+  std::swap(pending_media_file_creation_jobs_, empty_queue);
+
+  prefs_->ClearPref(kPlaylistItems);
+
+  NotifyPlaylistChanged(
+      {PlaylistChangeParams::ChangeType::CHANGE_TYPE_ALL_DELETED, ""});
+
+  CleanUp();
+}
+
+void PlaylistService::AddObserver(PlaylistServiceObserver* observer) {
+  observers_.AddObserver(observer);
+}
+
+void PlaylistService::RemoveObserver(
+    PlaylistServiceObserver* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void PlaylistService::OnMediaFileReady(base::Value&& playlist_value,
+                                       bool partial) {
+  // TODO(simonhong): Revisit how |partial| handles.
+  if (video_media_file_downloader_->in_progress() ||
+      audio_media_file_downloader_->in_progress())
+    partial = true;
+  VLOG(2) << __func__ << ": "
+          << *playlist_value.FindStringKey(kPlaylistPlaylistNameKey) << " "
+          << partial;
+
+  const std::string playlist_id =
+      *playlist_value.FindStringKey(kPlaylistIDKey);
+  UpdatePlaylistValue(playlist_id, std::move(playlist_value));
+
+  NotifyPlaylistChanged(
+      {partial
+           ? PlaylistChangeParams::ChangeType::CHANGE_TYPE_PLAY_READY_PARTIAL
+           : PlaylistChangeParams::ChangeType::CHANGE_TYPE_PLAY_READY,
+       playlist_id});
+
+  // TODO(simonhong): Revisit how |partial| handles.
+  // If partial is true here, one of video or audio media files are not ready.
+  if (partial)
+    return;
+
+  GenerateIndexHTMLFile(GetPlaylistItemDirPath(playlist_id));
+
+  if (!pending_media_file_creation_jobs_.empty())
+    GenerateMediaFiles();
+}
+
+void PlaylistService::OnMediaFileGenerationFailed(
+    base::Value&& playlist_value) {
+  VLOG(2) << __func__ << ": "
+          << *playlist_value.FindStringKey(kPlaylistPlaylistNameKey);
+
+  video_media_file_downloader_->RequestCancelCurrentPlaylistGeneration();
+  audio_media_file_downloader_->RequestCancelCurrentPlaylistGeneration();
+  const std::string playlist_id =
+      *playlist_value.FindStringKey(kPlaylistIDKey);
+  NotifyPlaylistChanged(
+      {PlaylistChangeParams::ChangeType::CHANGE_TYPE_ABORTED, playlist_id});
+
+  if (!pending_media_file_creation_jobs_.empty())
+    GenerateMediaFiles();
+}
+
+void PlaylistService::OnGetOrphanedPaths(
+    const std::vector<base::FilePath> orphaned_paths) {
+  if (orphaned_paths.empty()) {
+    VLOG(2) << __func__ << ": No orphaned playlist";
+    return;
+  }
+
+  for (const auto& path : orphaned_paths) {
+    VLOG(2) << __func__ << ": " << path << " is orphaned";
+    io_task_runner()->PostTask(FROM_HERE, base::BindOnce(&DeleteDir, path));
+  }
+}
+
+void PlaylistService::CleanUp() {
+  base::Value playlist = GetAllPlaylistItems();
+
+  base::flat_set<std::string> ids;
+  for (const auto& item : playlist.GetList()) {
+    const std::string* id = item.FindStringKey(kPlaylistIDKey);
```

Comment:
what happens if there is no ID? Can we add a comment?

Reply:
DCHECK added. This will not be happened. All playlist item must have it's id.